salt/requisite.py

```diff
@@ -0,0 +1,412 @@
+"""
+The classes and functions in this module are related to requisite
+handling and ordering of chunks by the State compiler
+"""
+
+
+from collections import defaultdict
+from collections.abc import Generator, Iterable, Sequence
+from enum import Enum, auto
+import fnmatch
+import logging
+from typing import Any, Optional, Union
+
+import networkx as nx  # pylint: disable=3rd-party-module-not-gated
+
+log = logging.getLogger(__name__)
+
+# See https://docs.saltproject.io/en/latest/ref/states/layers.html for details on the naming
+LowChunk = dict[str, Any]
```

Comment:
This works in python 3.8 if I add `from __future__ import annotations`?

Reply:
> This works in python 3.8 if I add `from __future__ import annotations`?

You need the future import and the right syntax. It should be `:` not `=`.

Reply:
I did some more research on this and could not find the documentation about this. All the documentation I could find shows that you can use `TypeAlias` `from typing` or in python 3.12 you can use the new `type` keyword. I did not use `TypeAlias`  because it was not added until python 3.10 and salt needs to support python 3.8. Can you provide some links to documentation on this?

Reply:
https://docs.python.org/3.8/library/__future__.html which links you to https://peps.python.org/pep-0563/

Wrong syntax and no futures import.

```
❯ python3.8 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 3, in <module>
    LowChunk = dict[str, Any]
TypeError: 'type' object is not subscriptable
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ python3.9 /tmp/foo.py
dict[str, typing.Any]
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)
```
```
❯ cat /tmp/foo.py
from typing import Any

LowChunk = dict[str, Any]

print(LowChunk)
```

Right syntax, you are defining a type, not a variable, no futures import.
```
❯ python3.8 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 3, in <module>
    LowChunk: dict[str, Any]
TypeError: 'type' object is not subscriptable
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ python3.9 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 5, in <module>
    print(LowChunk)
NameError: name 'LowChunk' is not defined
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
The above is actually good, because we defined the type, not a variable.
```
❯ cat /tmp/foo.py
from typing import Any

LowChunk: dict[str, Any]

print(LowChunk)
```

Right syntax, with the futures import:
```
❯ python3.8 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 7, in <module>
    print(LowChunk)
NameError: name 'LowChunk' is not defined
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ python3.9 /tmp/foo.py
Traceback (most recent call last):
  File "/tmp/foo.py", line 7, in <module>
    print(LowChunk)
NameError: name 'LowChunk' is not defined
~/projects/SaltStack/salt/hotfix/caches (hotfix/caches|✔)  exitcode: 1
```
```
❯ cat /tmp/foo.py
from __future__ import annotations

from typing import Any

LowChunk: dict[str, Any]

print(LowChunk)
```

Reply:
Is there something I need to do to make type checkers like pyright and mypy work with this syntax?